File: api-ts/src/utils/captureRedaction.ts
/**
 * Secret screening for free-text capture destined for the fine-tune dataset
 * (user_intents.raw_text). This is a SEPARATE control from Sentry's
 * `redactSecretsInText` (../lib/sentryRedact.ts) — Sentry redacts secrets
 * IN PLACE for diagnosability across arbitrary error payloads, so its hex
 * pattern deliberately catches anything >=40 hex chars (including plain
 * 20-byte EVM addresses, which are fine to log to Sentry as diagnostic
 * context). Capture is different: a wallet-address-containing trading
 * sentence ("swap to 0x1234...") is exactly the kind of normal input this
 * dataset needs, so a 40-char EVM address must NOT trip this screen. We
 * reuse Sentry's JWT / AWS-key patterns (unambiguous secret shapes) but keep
 * our own key-length-anchored (64 hex) and base58 (64-128) checks instead of
 * Sentry's broader >=40 hex rule.
 */
import { JWT_PATTERN, AWS_ACCESS_KEY_PATTERN } from '../lib/sentryRedact'

export interface SecretScreenResult {
	unsafe: boolean
	reason?: string
}

// BIP-39 wordlist is 2048 words; we don't ship the full list here (capture is
// a heuristic gate, not a wallet-import validator). Instead we detect the
// SHAPE of a mnemonic: N consecutive lowercase alphabetic words (each
// 3-8 chars, space separated) where N is a valid BIP-39 length. This is
// intentionally permissive — false positives just withhold text, which is
// the safe failure mode.
const MNEMONIC_LENGTHS = new Set([12, 15, 18, 21, 24])
const WORD_RE = /^[a-z]{3,8}$/

// Minimum length of a consecutive run of wordlike tokens that is treated as
// an embedded mnemonic even when it is NOT the entire message (e.g. "here is
// my seed phrase: <12 words>"). Ported from bot/utils/capture_redaction.py's
// `_MIN_CONSECUTIVE_WORDLIKE_TOKENS` fallback so the two stacks agree.
const MIN_CONSECUTIVE_WORDLIKE_TOKENS = 11

function looksLikeMnemonic(text: string): boolean {
	const words = text.trim().split(/\s+/).filter(Boolean)
	const lowered = words.map((w) => w.toLowerCase())

	// Whole-message exact-length mnemonic (all tokens wordlike).
	if (MNEMONIC_LENGTHS.has(lowered.length) && lowered.every((w) => WORD_RE.test(w))) {
		return true
	}

	// Consecutive-run fallback: catches a mnemonic embedded in a longer
	// sentence ("here is my seed phrase: <12 words>").
	let run = 0
	let best = 0
	for (const w of lowered) {
		if (WORD_RE.test(w)) {
			run += 1
			best = Math.max(best, run)
		} else {
			run = 0
		}
	}
	return best >= MIN_CONSECUTIVE_WORDLIKE_TOKENS
}

// Raw private key: optional 0x prefix + 64+ hex chars, not part of a longer
// run. Uses lookaround (not \b) so a 128-hex-char hex-encoded ed25519/Solana
// keypair — which \b cannot match mid-run of a longer hex string — is still
// caught. {64,} (not exactly 64) covers 64, 66 (0x+64), 128, and any longer
// hex secret. A 40-hex-char EVM/Tron address is well under the 64-char floor
// and stays unflagged.
const RAW_PRIVATE_KEY_RE = /(?<![0-9a-fA-F])(0x)?[0-9a-fA-F]{64,}(?![0-9a-fA-F])/

// Base58 secret-shaped run (Solana keypairs, WIF-adjacent blobs): 64-128 chars,
// base58 alphabet excludes 0/O/I/l.
const BASE58_SECRET_RE = /[1-9A-HJ-NP-Za-km-z]{64,128}/

function shannonEntropy(s: string): number {
	const counts = new Map<string, number>()
	for (const ch of s) counts.set(ch, (counts.get(ch) ?? 0) + 1)
	let entropy = 0
	for (const count of counts.values()) {
		const p = count / s.length
		entropy -= p * Math.log2(p)
	}
	return entropy
}

// Generic high-entropy token: a single contiguous non-whitespace run of
// >=32 chars with entropy > 3.5 bits/char (catches API keys, JWTs, base64
// secrets that don't match a more specific pattern above).
const TOKEN_RE = /\S{32,}/g

// Pure hex runs (with optional 0x prefix) are handled exclusively by
// RAW_PRIVATE_KEY_RE above (anchored at 64+ chars). A random hex
// string inherently carries ~4 bits/char of entropy, so without this
// exclusion any ordinary EVM address, tx hash, or hex id longer than 32
// chars would trip the generic entropy net as a false positive. Only
// non-hex tokens (base64, mixed-case alphanumeric secrets, etc.) fall
// through to the entropy check.
const HEX_TOKEN_RE = /^(0x)?[0-9a-fA-F]+$/

function hasHighEntropyToken(text: string): boolean {
	const matches = text.match(TOKEN_RE)
	if (!matches) return false
	return matches.some((tok) => !HEX_TOKEN_RE.test(tok) && shannonEntropy(tok) > 3.5)
}

/**
 * Screen free text before it is persisted to the append-only capture tables.
 * Fails closed: on any internal error, treat the text as unsafe.
 */
export function screenForSecrets(text: string): SecretScreenResult {
	try {
		if (!text) return { unsafe: false }

		if (looksLikeMnemonic(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}

		if (RAW_PRIVATE_KEY_RE.test(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}

		if (BASE58_SECRET_RE.test(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}

		// Reuse Sentry's unambiguous secret-value patterns (JWTs, AWS keys). Both
		// are exported with the 'g' flag for Sentry's .replace() use, so build a
		// fresh non-global RegExp per check here — reusing a global regex's
		// .test() across calls would carry stateful lastIndex bugs.
		if (new RegExp(JWT_PATTERN.source).test(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}
		if (new RegExp(AWS_ACCESS_KEY_PATTERN.source).test(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}

		if (hasHighEntropyToken(text)) {
			return { unsafe: true, reason: 'secret_detected' }
		}

		return { unsafe: false }
	} catch {
		// Fail closed: an unexpected error while screening must never result in
		// storing unscreened text.
		return { unsafe: true, reason: 'secret_detected' }
	}
}

File: bot/utils/capture_redaction.py
"""Secret screening for the data-capture pipeline (bot/services/capture_service.py).

FAIL CLOSED: on any ambiguity or internal error, treat input as unsafe. This
module gates whether raw user text is ever persisted to `user_intents` for
future fine-tuning — a false negative here means a private key or seed
phrase lands in a training dataset. A false positive just costs one row of
training text.

No vendored BIP-39 wordlist was found in this repo (checked eth_account,
searched for mnemonic/bip_utils packages and any vendored english.txt — none
present), so mnemonic detection falls back to a structural heuristic: a run
of >= 11 lowercase alpha tokens, each 3-8 chars, is treated as a probable
seed phrase. This is deliberately looser than word-list matching (some
ordinary sentences could trip it) because false positives are cheap and
false negatives are not.

Reuses the secret-shaped value patterns from bot/services/sentry_service.py
(`_SECRET_VALUE_PATTERNS`) rather than duplicating the hex/base58/JWT/AWS-key
regexes.
"""

from __future__ import annotations

import math
import re
from typing import Optional, Tuple

try:
    # Reuse the existing, audited secret-shape patterns instead of duplicating
    # them. These are private (leading underscore) but this module lives in
    # the same codebase and treats them as an internal shared constant.
    #
    # NOTE: sentry_service's catch-all long-hex pattern (`(0x)?[a-fA-F0-9]{40,}`)
    # is deliberately EXCLUDED here — a 40-hex-char EVM address matches it, and
    # addresses are explicitly not secrets for capture purposes (a swap
    # "send 1 eth to 0xabc...def" sentence must stay safe). Our own
    # `_HEX_PRIVATE_KEY` pattern below covers the 64-hex private-key case
    # instead. The Telegram-token / JWT / AWS-key / base58 patterns are
    # unambiguous and are reused as-is.
    from bot.services.sentry_service import _SECRET_VALUE_PATTERNS as _SENTRY_ALL_PATTERNS

    _SENTRY_SECRET_PATTERNS = tuple(p for p in _SENTRY_ALL_PATTERNS if "40," not in p.pattern)
except Exception:  # noqa: BLE001 — fail closed, not fail silent-import
    _SENTRY_SECRET_PATTERNS = ()

# --- Raw private key: 0x-prefixed or bare hex, 64+ chars ---
# Uses lookaround (not \b) so a 128-hex-char hex-encoded ed25519/Solana
# keypair — which \b cannot match mid-run of a longer hex string — is still
# caught. {64,} (not exactly 64) covers 64, 66 (0x+64), 128, and any longer
# hex secret. A 40-hex-char EVM/Tron address is well under the 64-char floor
# and stays unflagged.
_HEX_PRIVATE_KEY = re.compile(r"(?<![0-9a-fA-F])(0x)?[0-9a-fA-F]{64,}(?![0-9a-fA-F])")

# --- Solana / base58 secret keys (~64-128 base58 chars) ---
# Base58 alphabet excludes 0, O, I, l.
_BASE58_BLOB = re.compile(r"\b[1-9A-HJ-NP-Za-km-z]{64,128}\b")

# --- High-entropy whitespace-free token ---
_MIN_ENTROPY_LEN = 32
_MIN_ENTROPY_BITS = 3.5

# --- BIP-39-shaped mnemonic (structural fallback: no wordlist vendored) ---
_MNEMONIC_LENGTHS = {12, 15, 18, 21, 24}
_MNEMONIC_WORD = re.compile(r"^[a-z]{3,8}$")
_MIN_CONSECUTIVE_WORDLIKE_TOKENS = 11

# Attempt to load a real BIP-39 wordlist if one is vendored anywhere in the
# dependency tree. None found as of writing (see module docstring), but this
# keeps detection accurate for free if one shows up later (e.g. via
# `mnemonic` or `bip_utils` being added as a dependency).
_BIP39_WORDLIST: Optional[frozenset] = None
try:
    from mnemonic import Mnemonic as _Mnemonic  # type: ignore

    _BIP39_WORDLIST = frozenset(_Mnemonic("english").wordlist)
except Exception:  # noqa: BLE001
    try:
        from bip_utils import Bip39WordsListGetter, Bip39Languages  # type: ignore

        _wordlist_obj = Bip39WordsListGetter().GetByLanguage(Bip39Languages.ENGLISH)
        _BIP39_WORDLIST = frozenset(str(w) for w in _wordlist_obj.GetWords())
    except Exception:  # noqa: BLE001
        _BIP39_WORDLIST = None

_MIN_WORDLIST_HITS = 8


def _shannon_entropy(token: str) -> float:
    if not token:
        return 0.0
    freq: dict[str, int] = {}
    for ch in token:
        freq[ch] = freq.get(ch, 0) + 1
    length = len(token)
    entropy = 0.0
    for count in freq.values():
        p = count / length
        entropy -= p * math.log2(p)
    return entropy


def _looks_like_mnemonic(tokens: list[str]) -> bool:
    """Detect a BIP-39-shaped mnemonic among whitespace-split tokens."""
    lowered = [t.lower() for t in tokens if t]

    if _BIP39_WORDLIST is not None:
        # Real wordlist available: look for >= _MIN_WORDLIST_HITS consecutive
        # tokens that are wordlist members.
        run = 0
        for tok in lowered:
            if tok in _BIP39_WORDLIST:
                run += 1
                if run >= _MIN_WORDLIST_HITS:
                    return True
            else:
                run = 0
        return False

    # Fallback heuristic: no wordlist available. Treat >= 11 consecutive
    # alpha tokens, all lowercase, length 3-8, as unsafe. This also matches
    # phrases of the canonical 12/15/18/21/24 lengths when the whole input
    # is (close to) just the phrase.
    run = 0
    best = 0
    for tok in lowered:
        if _MNEMONIC_WORD.match(tok):
            run += 1
            best = max(best, run)
        else:
            run = 0
    if best >= _MIN_CONSECUTIVE_WORDLIKE_TOKENS:
        return True

    # Exact canonical mnemonic length + every token wordlike -> unsafe even
    # if the run counter above didn't fire (e.g. exactly 12 tokens total).
    if len(lowered) in _MNEMONIC_LENGTHS and all(_MNEMONIC_WORD.match(t) for t in lowered):
        return True

    return False


def screen_for_secrets(text: str) -> Tuple[bool, Optional[str]]:
    """Screen free text for secrets before it is persisted for training data.

    Returns (is_unsafe, reason). Fails closed: any internal error or
    ambiguity is treated as unsafe with reason 'screen_error'.

    This function must never raise.
    """
    try:
        if text is None:
            return False, None
        if not isinstance(text, str):
            return True, "non_string_input"

        stripped = text.strip()
        if not stripped:
            return False, None

        tokens = stripped.split()

        # 1. Mnemonic-shaped input.
        if _looks_like_mnemonic(tokens):
            return True, "mnemonic_detected"

        # 2. Raw private key (hex).
        if _HEX_PRIVATE_KEY.search(stripped):
            return True, "private_key_detected"

        # 3. Solana/base58 secret key.
        if _BASE58_BLOB.search(stripped):
            return True, "base58_secret_detected"

        # 4. Reused sentry secret-value patterns (Telegram token, JWT, AWS key,
        #    long hex, base58 — kept in sync with sentry_service.py).
        for pattern in _SENTRY_SECRET_PATTERNS:
            if pattern.search(stripped):
                return True, "secret_pattern_detected"

        # 5. High-entropy whitespace-free token. Pure hex tokens (e.g. a
        # 40-hex-char EVM/Tron address) are excluded here: hex has a fixed
        # max entropy of 4 bits/char, so essentially ANY hex string of this
        # length trips the threshold, and addresses are explicitly not
        # secrets. Actual hex private keys are still caught by
        # `_HEX_PRIVATE_KEY` above (64 hex chars).
        for token in tokens:
            if len(token) < _MIN_ENTROPY_LEN:
                continue
            bare = token[2:] if token.lower().startswith("0x") else token
            if re.fullmatch(r"[0-9a-fA-F]+", bare):
                continue
            if _shannon_entropy(token) > _MIN_ENTROPY_BITS:
                return True, "high_entropy_token"

        return False, None
    except Exception:  # noqa: BLE001 — fail closed on any unexpected error
        return True, "screen_error"
